addons/l10n_se_ocr/models/account_move.py

```diff
@@ -0,0 +1,56 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import api, fields, models, _
+from odoo.exceptions import UserError
+from odoo.addons.l10n_se.util.luhn import luhn
+
+class AccountMove(models.Model):
+    _inherit = 'account.move'
+
+    def _get_invoice_reference_se_ocr2_invoice(self):
+        self.ensure_one()
+
+        number = str(self.id)
+        reference = luhn.append(number)
+
+        return reference
+
+    def _get_invoice_reference_se_ocr3_invoice(self):
+        self.ensure_one()
+        number = str(self.id)
+
+        number = number + str(len(number)+2)[:1]
+        reference = luhn.append(number)
+
+        return reference
+
+    def _get_invoice_reference_se_ocr4_invoice(self):
+        self.ensure_one()
+        number = str(self.id)
+
+        ocr_length = self.journal_id.invoice_ocr_length
+
+        if ocr_length < 5:
+            raise UserError(_("OCR Reference Number length need to be greater then 5. Please correct settings under invoice journal settings."))
+
+        if len(number)+1 > ocr_length:
+            raise UserError(_("OCR Reference Number length is greater than allowed. Allowed length in invoice journal setting is %s.") % str(ocr_length))
+
+        number = number.ljust(ocr_length-1, '0')
+        reference = luhn.append(number)
+
+        return reference
+
+    @api.onchange('partner_id')
+    def onchange_partner_id(self):
```

Comment:
Maybe better as a private method and calling super as well

Reply:
Made private in next update.